### ISSUE TITLE
 build kernel with GPU support failed

### ISSUE THREAD

**Issue by fighterhit** [kata-containers/kata-containers#3380]
When I follow the [document](https://github.com/kata-containers/kata-containers/blob/main/docs/use-cases/Nvidia-GPU-passthrough-and-Kata.md#build-kata-containers-kernel-with-gpu-support) to build a kernel with GPU support, the error is reported as follows :

- my command:
```
 ./build-kernel.sh -v 4.19.86 -g nvidia -f setup -d
```

- results:
```
INFO: Config version: 88
INFO: Kernel version: 4.19.86
INFO: kernel path does not exist, will download kernel
INFO: Download kernel version 4.19.86
INFO: Download kernel
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   162  100   162    0     0    329      0 --:--:-- --:--:-- --:--:--   329
100 98.5M  100 98.5M    0     0  2957k      0  0:00:34  0:00:34 --:--:-- 14.8M
linux-4.19.86.tar.xz: OK
INFO: Apply patches from /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/patches/4.19.x
INFO: Found 6 patches
INFO: Apply /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/patches/4.19.x/0001-4.19.x-PTP_KVM-support-for-arm-arm64.patch
INFO: Apply /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/patches/4.19.x/0002-Enable-memory-hotplug-using-probe-for-arm64.patch
INFO: Apply /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/patches/4.19.x/0003-NO-UPSTREAM-9P-always-use-cached-inode-to-fill-in-v9.patch
INFO: Apply /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/patches/4.19.x/0004-Compile-in-evged-always.patch
INFO: Apply /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/patches/4.19.x/0005-arm64-backport-Arm64-KVM-Dynamic-IPA-and-52bit-IPA-s.patch
INFO: Apply /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/patches/4.19.x/0006-net-virtio_vsock-Fix-race-condition-between-bind-and.patch
INFO: Add kernel config for GPU due to '-g nvidia'
INFO: Constructing config from fragments: /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/configs/fragments/x86_64/.config
INFO: Some CONFIG elements failed to make the final .config:
INFO: Value requested for CONFIG_FUSE_DAX not in final .config
Value requested for CONFIG_VIRTIO_FS not in final .config
Value requested for CONFIG_VIRTIO_PMEM not in final .config
Value requested for CONFIG_PVH not in final .config
Value requested for CONFIG_VIRTIO_MEM not in final .config
INFO: Generated config file can be found in /home/fighter/Projects/golang/kata-containers/tools/packaging/kernel/configs/fragments/x86_64/.config
ERROR: Failed to construct requested .config file
ERROR: failed to find default config 
```

Sorry, I am not very familiar with the kernel, so I am not sure if I am missing something to configure. In addition, I think that the use of GPU in the document is more complicated. Does the community consider providing a GPU-supporting guest kernel? Thanks!

**Comment by Bevisy** [comment]
I tried to use kernel v5.10.25(latest version) and it worked normally
```sh
❯ ./build-kernel.sh -v 5.10.25 -g nvidia -f setup -d
INFO: Config version: 88
INFO: Kernel version: 5.10.25
INFO: /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/kata-linux-5.10.25-88 already exist
INFO: Force generate config due to '-f'
INFO: Apply patches from /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x
INFO: Found 4 patches
INFO: Apply /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x/0001-5.10.x-PTP_KVM-support-for-arm-arm64.patch
INFO: Apply /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x/0001-NO-UPSTREAM-9P-always-use-cached-inode-to-fill-in-v9.patch
INFO: Apply /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x/0001-arm64-mmu-compared-with-linear-start-physical-addres.patch
INFO: Apply /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x/0001-arm64-sparsemem-reduce-SECTION_SIZE_BITS.patch
INFO: Add kernel config for GPU due to '-g nvidia'
INFO: Constructing config from fragments: /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/configs/fragments/x86_64/.config
INFO: Copying config file from: /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/configs/fragments/x86_64/.config
#
# configuration written to .config
#
Kernel source ready: /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/kata-linux-5.10.25-88 
```
But when using v4.19.86, there are something wrong like what you did. You can try changing the kernel version, it won't block you from the next verification step.

**Comment by fighterhit** [comment]
> I tried to use kernel v5.10.25(latest version) and it worked normally
> 
> ```shell
> ❯ ./build-kernel.sh -v 5.10.25 -g nvidia -f setup -d
> INFO: Config version: 88
> INFO: Kernel version: 5.10.25
> INFO: /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/kata-linux-5.10.25-88 already exist
> INFO: Force generate config due to '-f'
> INFO: Apply patches from /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x
> INFO: Found 4 patches
> INFO: Apply /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x/0001-5.10.x-PTP_KVM-support-for-arm-arm64.patch
> INFO: Apply /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x/0001-NO-UPSTREAM-9P-always-use-cached-inode-to-fill-in-v9.patch
> INFO: Apply /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x/0001-arm64-mmu-compared-with-linear-start-physical-addres.patch
> INFO: Apply /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/patches/5.10.x/0001-arm64-sparsemem-reduce-SECTION_SIZE_BITS.patch
> INFO: Add kernel config for GPU due to '-g nvidia'
> INFO: Constructing config from fragments: /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/configs/fragments/x86_64/.config
> INFO: Copying config file from: /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/configs/fragments/x86_64/.config
> #
> # configuration written to .config
> #
> Kernel source ready: /home/zbb/go/src/github.com/kata-containers/kata-containers/tools/packaging/kernel/kata-linux-5.10.25-88 
> ```
> 
> But when using v4.19.86, there are something wrong like what you did. You can try changing the kernel version, it won't block you from the next verification step. In fact, I verified that the way the documentation provides support for the GPU no longer seems to work.

Thank you @Bevisy , yes I have tried the v5.10.25 version as you did and it does work, and I also use the GPU passthrough method to pass the GPU into the container according to the documentation. But when I install the NVIDIA GPU driver in the container, I keep getting the following error. Have you encountered this problem?

```
ERROR: Unable to load the kernel module 'nvidia.ko'.  This happens most frequently when this kernel module was built against the wrong or improperly configured kernel sources, with a version of gcc that differs from the one used to build the target kernel, or if another driver, such as nouveau, is present and prevents the NVIDIA kernel module from obtaining ownership of the NVIDIA GPU(s), or no NVIDIA GPU installed in this system is supported by this NVIDIA Linux graphics driver release.

Please see the log entries 'Kernel module load error' and 'Kernel messages' at the end of the file '/var/log/nvidia-installer.log' for more information.
-> Kernel module load error: No such device
-> Kernel messages:
[    3.856188] pci 0000:01:04.0: PME# supported from D0 D3hot
[    3.856430] pci 0000:01:04.0: 0.000 Gb/s available PCIe bandwidth, limited by Unknown x0 link at 0000:00:02.0 (capable of 126.016 Gb/s with 8.0 GT/s PCIe x16 link)
[    3.856664] pci 0000:01:04.0: BAR 0: assigned [mem 0xfe404000-0xfe404fff]
[    3.863210] pci 0000:00:02.0: PCI bridge to [bus 01]
[    3.863228] pci 0000:00:02.0:   bridge window [io  0xc000-0xcfff]
[    3.864444] pci 0000:00:02.0:   bridge window [mem 0xfe400000-0xfe5fffff]
[    3.865235] pci 0000:00:02.0:   bridge window [mem 0xfe800000-0xfe9fffff 64bit pref]
[    3.886543] kmem.limit_in_bytes is deprecated and will be removed. Please report your usecase to linux-mm@kvack.org if you depend on this functionality.
[    3.895363] cgroup: cgroup: disabling cgroup2 socket matching due to net_prio or net_cls activation
[    3.910477] systemd[1]: run-libcontainer-6222bf92370104fc5d76c6ca677e7e5bc1702f31310198426b5fb7d1ea4c0d21-6222bf92370104fc5d76c6ca677e7e5bc1702f31310198426b5fb7d1ea4c0d21-runc.Y2Ke8b.mount: Succeeded.
[  501.752520] systemd[1]: run-libcontainer-6222bf92370104fc5d76c6ca677e7e5bc1702f31310198426b5fb7d1ea4c0d21-6222bf92370104fc5d76c6ca677e7e5bc1702f31310198426b5fb7d1ea4c0d21-runc.BnR1mb.mount: Succeeded.
[  848.513080] nvidia: loading out-of-tree module taints kernel.
[  848.513089] nvidia: module license 'NVIDIA' taints kernel.
[  848.513090] Disabling lock debugging due to kernel taint
[  848.534328] nvidia: module verification failed: signature and/or required key missing - tainting kernel
[  848.547065] nvidia-nvlink: Nvlink Core is being initialized, major device number 245

[  848.563654] nvidia 0000:01:01.0: enabling device (0100 -> 0101)
[  848.579445] NVRM: request_mem_region failed for 0M @ 0x0. This can
[  848.579445] NVRM: occur when a driver such as rivatv is loaded and claims
[  848.579445] NVRM: ownership of the device's registers.
[  848.579453] nvidia: probe of 0000:01:01.0 failed with error -1
[  848.579475] NVRM: The NVIDIA probe routine failed for 1 device(s).
[  848.579476] NVRM: None of the NVIDIA devices were initialized.
[  848.579912] nvidia-nvlink: Unregistered the Nvlink Core, major device number 245
ERROR: Installation has failed.  Please see the file '/var/log/nvidia-installer.log' for details.  You may find suggestions on fixing installation problems in the README available on the Linux driver download page at www.nvidia.com.
```

**Comment by Bevisy** [comment]
Not exactly the same, but also the driver installation failed.

**Comment by fighterhit** [comment]
I think we need some help. ping @Jimmy-Xu

**Comment by fighterhit** [comment]
By the way, when using the v5.10.25 kernel, I found that when I create a debian container, `apt update` will fail with the following errors. 
```
Ign:1 http://deb.debian.org/debian stretch InRelease
Get:2 http://security.debian.org/debian-security stretch/updates InRelease [53.0 kB]
Get:3 http://deb.debian.org/debian stretch-updates InRelease [93.6 kB]
Get:4 http://deb.debian.org/debian stretch Release [118 kB]     
Get:5 http://deb.debian.org/debian stretch Release.gpg [3177 B]
Get:6 http://deb.debian.org/debian stretch/main amd64 Packages [7080 kB]                                                                                                                                       
Reading package lists... Done                                                                                                                                                                                  
E: Unable to determine file size for fd 9 - fstat (2: No such file or directory)
E: The repository 'http://security.debian.org/debian-security stretch/updates InRelease' provides only weak security information.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
E: Unable to determine file size for fd 9 - fstat (2: No such file or directory)
E: The repository 'http://deb.debian.org/debian stretch-updates InRelease' provides only weak security information.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```

I fix it according to this [issue](https://github.com/kata-containers/documentation/issues/776), in which the patch below does not actually take effect. So I have to manually modify this part of the kernel code and I think the community may need to pay attention to this. @jodh-intel 

https://github.com/kata-containers/kata-containers/blob/0bd2cc5a93edfc5a59aa7a7409dbeaac494cd038/tools/packaging/kernel/patches/5.10.x/0001-NO-UPSTREAM-9P-always-use-cached-inode-to-fill-in-v9.patch#L21-L45

**Comment by jodh-intel** [comment]
@fighterhit - What version of Kata are you using here and what config? You seem to have hit the long-standing "9p unlink stat" bug which has been "fixed" multiple times fwicr. However, the latest official upstream fix only appears in Linux v5.11:

- https://github.com/torvalds/linux/commits/154372e67d4053e56591245eb413686621941333

The default for Kata 2.x is `virtio-fs` as 9p has always been problematic for this and other reasons.

However, if you are using an old environment you can avoid hacking your guest kernel by using use the `ramfs`/`tmpfs` hack as shown on:

- https://github.com/intel/cc-oci-runtime/issues/152#issuecomment-241728454

Reference: https://github.com/intel/cc-oci-runtime/issues/47.

/cc @gkurz

**Comment by fighterhit** [comment]
> @fighterhit - What version of Kata are you using here and what config? You seem to have hit the long-standing "9p unlink stat" bug which has been "fixed" multiple times fwicr. However, the latest official upstream fix only appears in Linux v5.11:
> 
> * https://github.com/torvalds/linux/commits/154372e67d4053e56591245eb413686621941333
> 
> The default for Kata 2.x is `virtio-fs` as 9p has always been problematic for this and other reasons.
> 
> However, if you are using an old environment you can avoid hacking your guest kernel by using use the `ramfs`/`tmpfs` hack as shown on:
> 
> * [networkstatic/iperf3 containers fails to run with cors intel/cc-oci-runtime#152 (comment)](https://github.com/intel/cc-oci-runtime/issues/152#issuecomment-241728454)
> 
> Reference: [intel/cc-oci-runtime#47](https://github.com/intel/cc-oci-runtime/issues/47).
> 
> /cc @gkurz

Hi @jodh-intel, thanks for your reply. For some reason, in order to test the GPU via docker, I installed the 1.12 version of `kata-runtime` on the host via `snap`. But I built the guest-kernel through the main branch of the source code. Is this problem related to the guest-kernel or the 1.12 version of kata-runtime?

**Comment by jodh-intel** [comment]
@fighterhit - If you are sure you were using the kernel you built, the problem would be your guest kernel version. But Kata 1.x is no longer being maintained so we would strongly recommend that you use Kata 2.x for all use-cases.

---

I'm also unclear why your snap installation is so old since snap is supposed to automatically update your installation as new releases become available. The last stable Kata 1.x release was 1.12.1 but the latest snap version is 2.3.0:

- https://snapcraft.io/kata-containers

**Comment by fighterhit** [comment]
Thank you for the explanation @jodh-intel . If I remember correctly, when I installed Kata via snap, the version was specified as 1.12, and it did keep this version unupdated until now. My host system is debian stretch.

Since our current environment is Docker(19.03.12)+K8S(<=1.18), I had to use Kata 1.x for testing. But, if possible, I plan to use containerd directly in the future. In addition, I have paid attention to your recent explanation of this [issue](https://github.com/kata-containers/kata-containers/issues/3417#issuecomment-1009786515), and I am looking forward to the progress of the "wrapper".

---

Actually, I mainly want to manage NVIDIA GPU drivers via Kata. Because all containers on the same node in a k8s cluster use the same version of the NVIDIA driver, but sometimes some containers want to use a newer version, it is relatively difficult to maintain and upgrade the NVIDIA driver version. So I plan to install different versions of NVIDIA drivers in different kata virtual machines, and then make them base container images through `docker commit`. This way, when I create a container later, I can choose a different version of the NVIDIA driver. May I ask if this would work?
